File: uni_modules/ui-component/utssdk/app-ios/swift/observer/LiveStoreObserver.swift
import Combine
import DCloudUTSFoundation
import RTCRoomEngine_Plus

public class LiveStoreObserver {
    private var cancellables = Set<AnyCancellable>()
    public static let shared = LiveStoreObserver()

    public func liveStoreChanged(_ callback: @escaping (_ name: String, _ data: String) -> Void) {
        cancellables.removeAll()

        V2RTCRoomEngine.shared.getLiveStore()
            .state.subscribe(StateSelector(keyPath: \V2LiveState.liveList))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] liveList in
                guard let self = self else { return }
                var list = []
                for liveInfo in liveList {
                    list.append(convertLiveInfoToDic(liveInfo: liveInfo))
                }
                if let jsonList = JsonUtil.toJson(list) {
                    callback("liveList", jsonList)
                }
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getLiveStore()
            .state.subscribe(StateSelector(keyPath: \V2LiveState.liveListCursor))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] cursor in
                guard let self = self else { return }
                callback("liveListCursor", cursor)
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getLiveStore()
            .state.subscribe(StateSelector(keyPath: \V2LiveState.currentLive))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] liveInfo in
                guard let self = self else { return }
                if let liveInfo = liveInfo {
                    let dict = convertLiveInfoToDic(liveInfo: liveInfo)
                    if let json = JsonUtil.toJson(dict) {
                        callback("currentLive", json)
                    }
                } else {
                    callback("currentLive", "")
                }
            })
            .store(in: &cancellables)

        V2RTCRoomEngine.shared.getLiveStore()
            .state.subscribe(StateSelector(keyPath: \V2LiveState.localLiveStatus))
            .removeDuplicates()
            .receive(on: DispatchQueue.main)
            .sink(receiveValue: { [weak self] status in
                guard let self = self else { return }
                callback("localLiveStatus", String(status.rawValue))
            })
            .store(in: &cancellables)
    }

    private func convertLiveInfoToDic(liveInfo: LiveInfo) -> [String: Any] {
        var dict: [String: Any] = [
            "liveId": liveInfo.liveId,
            "liveName": liveInfo.liveName ?? "",
            "liveDescription": liveInfo.liveDescription ?? "",
            "categoryList": liveInfo.categoryList ?? [],
            "coverUrl": liveInfo.coverUrl ?? "",
            "backgroundUrl": liveInfo.backgroundUrl ?? "",
            "liveOwner": convertTUILoginUserToDic(userInfo: liveInfo.liveOwner),
            "currentViewerCount": liveInfo.currentViewerCount,
            "totalViewerCount": liveInfo.totalViewerCount,
            "createTime": liveInfo.createTime,
            "isMessageDisable": liveInfo.isMessageDisable,
            "isGiftEnabled": liveInfo.isGiftEnabled,
            "isPublicVisible": liveInfo.isPublicVisible,
            "isSeatEnabled": liveInfo.isSeatEnabled,
            "seatMode": convertSeatMode(liveInfo.seatMode),
            "maxSeatCount": liveInfo.maxSeatCount,
            "layoutTemplate": liveInfo.layoutTemplate,
            "customInfo": liveInfo.customInfo,
        ]
        return dict
    }
    private func convertSeatMode(_ mode: TUISeatMode) -> String {
        switch mode {
            case .freeToTake:
                return "FREE_TO_TAKE"
            case .applyToTake:
                return "APPLY_TO_TAKE"
            default:
                return ""
        }
    }
    private func convertTUILoginUserToDic(userInfo: TUILoginUserInfo?) -> [String: Any] {
        let dict: [String: Any] = [
            "userId": userInfo?.userId ?? "",
            "userName": userInfo?.userName ?? "",
            "avatarUrl": userInfo?.avatarUrl ?? "",
        ]
        return dict
    }
}

File: uni_modules/ui-component/utssdk/app-android/kotlin/observer/LiveStoreObserver.kt
package uts.sdk.modules.uiComponent.observer

import com.google.gson.Gson
import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine.SeatMode
import com.tencent.cloud.uikit.state.LiveInfo
import com.tencent.cloud.uikit.state.V2RTCRoomEngine
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

object LiveStoreObserver {
    private val gson = Gson()
    private var bindDataJob: Job? = null

    fun liveStoreChanged(callback: (String, String) -> Unit) {
        bindDataJob?.cancel()
        bindDataJob = CoroutineScope(Dispatchers.Main).launch {
            launch {
                V2RTCRoomEngine.getInstance().getLiveStore().liveState.liveList.collect { liveList ->
                    val list = liveList?.map { convertLiveInfoToMap(it) }
                    callback("liveList", gson.toJson(list))
                }
            }
            launch {
                V2RTCRoomEngine.getInstance().getLiveStore().liveState.liveListCursor.collect { cursor ->
                    callback("liveListCursor", gson.toJson(cursor))
                }
            }

            launch {
                V2RTCRoomEngine.getInstance().getLiveStore().liveState.currentLive.collect { liveInfo ->
                    val info = convertLiveInfoToMap(liveInfo)
                    callback("currentLive", gson.toJson(info))
                }
            }

            launch {
                V2RTCRoomEngine.getInstance().getLiveStore().liveState.localLiveStatus.collect { status ->
                    callback("localLiveStatus", gson.toJson(status.value))
                }
            }
        }
    }

    private fun convertLiveInfoToMap(liveInfo: LiveInfo): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        map["liveId"] = liveInfo.liveId
        map["liveName"] = liveInfo.liveName
        map["liveDescription"] = liveInfo.liveDescription
        map["categoryList"] = liveInfo.categoryList
        map["coverUrl"] = liveInfo.coverUrl
        map["backgroundUrl"] = liveInfo.backgroundUrl
        map["liveOwner"] = liveInfo.liveOwner
        map["currentViewerCount"] = liveInfo.currentViewerCount
        map["totalViewerCount"] = liveInfo.totalViewerCount
        map["createTime"] = liveInfo.createTime
        map["isMessageDisable"] = liveInfo.isMessageDisable
        map["isGiftEnabled"] = liveInfo.isGiftEnabled
        map["isPublicVisible"] = liveInfo.isPublicVisible
        map["keepOwnerOnSeat"] = liveInfo.keepOwnerOnSeat
        map["isSeatEnabled"] = liveInfo.isSeatEnabled
        map["seatMode"] = convertSeatMode(liveInfo.seatMode)
        map["maxSeatCount"] = liveInfo.maxSeatCount
        map["layoutTemplate"] = liveInfo.layoutTemplate
        map["customInfo"] = liveInfo.customInfo
        return map
    }

    private fun convertSeatMode(mode: SeatMode): String {
        return when (mode) {
            SeatMode.FREE_TO_TAKE -> "FREE_TO_TAKE"
            SeatMode.APPLY_TO_TAKE -> "APPLY_TO_TAKE"
            else -> ""
        }
    }
} 